nestkernel/recording_backend_mpi.cpp

```diff
@@ -0,0 +1,300 @@
+/*
+ *  recording_backend_mpi.cpp
+ *
+ *  This file is part of NEST.
+ *
+ *  Copyright (C) 2004 The NEST Initiative
+ *
+ *  NEST is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  NEST is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// C++ includes:
+#include <iostream>
+
+
+// Includes from nestkernel:
+#include "recording_device.h"
+#include "recording_backend_mpi.h"
+
+void
+nest::RecordingBackendMPI::initialize()
+{
+  auto nthreads = kernel().vp_manager.get_num_threads();
+  device_map devices( nthreads );
+  devices_.swap( devices );
+  comm_map comms( nthreads );
+  commMap_.swap( comms );
+}
+
+void
+nest::RecordingBackendMPI::finalize()
+{
+  // clear vector of map
+  for ( auto& it_device : devices_ )
+  {
+    it_device.clear();
+  }
+  for ( auto& it_comm : commMap_ )
+  {
+    it_comm.clear();
+  }
+  devices_.clear();
+  commMap_.clear();
+}
+
+void
+nest::RecordingBackendMPI::enroll( const RecordingDevice& device, const DictionaryDatum& params )
+{
+  if ( device.get_type() == RecordingDevice::SPIKE_DETECTOR )
+  {
+    thread tid = device.get_thread();
+    index node_id = device.get_node_id();
+
+    auto device_it = devices_[ tid ].find( node_id );
+    if ( device_it != devices_[ tid ].end() )
+    {
+      devices_[ tid ].erase( device_it );
+    }
+
+    std::pair< MPI_Comm*, const RecordingDevice* > pair = std::make_pair( nullptr, &device );
+    devices_[ tid ].insert( std::make_pair( node_id, pair ) );
+  }
+  else
+  {
+    throw BadProperty( "Only spike detectors can record to recording backend 'mpi'." );
+  }
+}
+
+void
+nest::RecordingBackendMPI::disenroll( const RecordingDevice& device )
+{
+  const auto tid = device.get_thread();
+  const auto node_id = device.get_node_id();
+
+  auto device_it = devices_[ tid ].find( node_id );
+  if ( device_it != devices_[ tid ].end() )
+  {
+    devices_[ tid ].erase( device_it );
+  }
+}
+
+void
+nest::RecordingBackendMPI::set_value_names( const RecordingDevice& device,
+  const std::vector< Name >& double_value_names,
+  const std::vector< Name >& long_value_names )
+{
+  // nothing to do
+}
+
+void
+nest::RecordingBackendMPI::prepare()
+{
+  // Create the connection with MPI per thread
+  // TODO validate approach
```

Comment:
Should this TODO be fixed?

Reply:
Dear @hakonsbm this TODO is a bit complicated to solve. We have asked @jougs for a meeting to discuss this issue.

Reply:
solved